trackpy/linking.py

```diff
@@ -735,146 +843,262 @@ def link_iter(levels, search_range, memory=0,
 
-    # Assume everything in first level starts a Track.
-    track_lst = [track_cls(p) for p in prev_set]
-    mem_set = set()
-
-    # Initialize memory with empty sets.
-    mem_history = []
-    for j in range(memory):
-        mem_history.append(set())
-
-    yield list(prev_set)  # Short-circuit the loop on first call.
-
-    for cur_level in levels:
-        # Create the set for the destination level.
-        cur_set = set(cur_level)
-        tmp_set = set(cur_level)  # copy used in next loop iteration
-
-        # First, a bit of unfinished business:
-        #
-        # If prediction is enabled, we need to update the positions in
-        # prev_hash to where we think they'll be in the frame corresponding to
-        # cur_level.
-        if predictor is not None:
-            # This only works for KDTree right now, because KDTree can store
-            # particle positions in a separate data structure from the PointND
-            # instances.
-            if not isinstance(prev_hash, TreeFinder):
-                raise NotImplementedError(
-                    'Prediction works with the "KDTree" neighbor_strategy only.')
-            # Get the time of cur_level from its first particle
-            t_next = list(itertools.islice(cur_level, 0, 1))[0].t
-            targeted_predictor = functools.partial(predictor, t_next)
-            prev_hash.rebuild(coord_map=targeted_predictor)  # Rewrite positions
-
-        # Now we can process the new particles.
-        # Make a Hash / Tree for the destination level.
-        if neighbor_strategy == 'BTree':
-            cur_hash = hash_generator()
+            # sort the candidate lists by distance
             for p in cur_set:
-                cur_hash.add_point(p)
-        elif neighbor_strategy == 'KDTree':
-            cur_hash = TreeFinder(cur_level)
-
-        # Set up attributes for keeping track of possible connections.
-        for p in cur_set:
-            p.back_cands = []
-            p.forward_cands = []
+                p.back_cands.sort(key=lambda x: x[1])
+            for p in prev_set:
+                p.forward_cands.sort(key=lambda x: x[1])
 
-        # Sort out what can go to what.
-        assign_candidates(cur_level, prev_hash, search_range,
-                          neighbor_strategy)
+            # Note that this modifies cur_set, prev_set, but that's OK.
+            spl, dpl = self._assign_links(cur_set, prev_set, self.search_range)
 
-        # sort the candidate lists by distance
-        for p in cur_set:
-            p.back_cands.sort(key=lambda x: x[1])
-        for p in prev_set:
-            p.forward_cands.sort(key=lambda x: x[1])
+            new_mem_set = set()
+            for sp, dp in zip(spl, dpl):
+                # Do linking
+                if sp is not None and dp is not None:
+                    sp.track.add_point(dp)
+                    if sp in self.mem_set:  # Very rare
+                        self.mem_set.remove(sp)
```

Comment:
@danielballan: The memory-removal logic from old line 871 ended up here during the adaptive search refactor.

I suppose I could move all the new stuff into a PR against the adaptive search branch of my repo. If you think it would make these diffs significantly less confusing, I'll do it.